x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/datafeed/MlRemoteLicenseChecker.java

```diff
@@ -0,0 +1,191 @@
+                    listener.onResponse(new LicenseViolation(
+                            new RemoteClusterLicenseInfo(clusterName.get(), xPackInfoResponse.getLicenseInfo())));
+                    return;
+                }
+
+                if (itr.hasNext()) {
+                    clusterName.set(itr.next());
+                    remoteClusterLicense(clusterName.get(), this);
+                } else {
+                    listener.onResponse(new LicenseViolation(null));
+                }
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                String message = "Could not determine the X-Pack licence type for cluster [" + clusterName.get() + "]";
+                if (e instanceof ActionNotFoundTransportException) {
+                    // This is likely to be because x-pack is not installed in the target cluster
+                    message += ". Is X-Pack installed on the target cluster?";
+                }
+                listener.onFailure(new ElasticsearchException(message, e));
+            }
+        };
+
+        remoteClusterLicense(clusterName.get(), infoListener);
+    }
+
+    private void remoteClusterLicense(String clusterName, ActionListener<XPackInfoResponse> listener) {
+        Client remoteClusterClient = client.getRemoteClusterClient(clusterName);
+        ThreadContext threadContext = remoteClusterClient.threadPool().getThreadContext();
+        try (ThreadContext.StoredContext ignore = threadContext.stashContext()) {
+            // we stash any context here since this is an internal execution and should not leak any
+            // existing context information.
+            threadContext.markAsSystemContext();
+
+            XPackInfoRequest request = new XPackInfoRequest();
+            request.setCategories(EnumSet.of(XPackInfoRequest.Category.LICENSE));
+            remoteClusterClient.execute(XPackInfoAction.INSTANCE, request, listener);
+        }
+    }
+
+    static boolean licenseSupportsML(XPackInfoResponse.LicenseInfo licenseInfo) {
+        License.OperationMode mode = License.OperationMode.resolve(licenseInfo.getMode());
+        return licenseInfo.getStatus() == License.Status.ACTIVE &&
+                (mode == License.OperationMode.PLATINUM || mode == License.OperationMode.TRIAL);
+    }
+
+    public static boolean isRemoteIndex(String index) {
+        return index.indexOf(':') != -1;
+    }
+
+    public static boolean containsRemoteIndex(List<String> indices) {
+        return indices.stream().anyMatch(MlRemoteLicenseChecker::isRemoteIndex);
+    }
+
+    /**
+     * Get any remote indices used in cross cluster search.
+     * Remote indices are of the form {@code cluster_name:index_name}
+     * @return List of remote cluster indices
+     */
+    public static List<String> remoteIndices(List<String> indices) {
+        return indices.stream().filter(index -> index.indexOf(':') != -1).collect(Collectors.toList());
```

Comment:
nit: The other methods use `filter(MlRemoteLicenseChecker::isRemoteIndex)` instead of repeating the low level predicate.